File: backend/nodejs-nest/src/app/auth/auth.service.ts
import { Injectable } from '@nestjs/common'
import { JwtService } from '@nestjs/jwt'
import { Role } from 'src/entities/role.enum'
import { User } from 'src/entities/user.entity'

import { UsersService } from '../users/users.service'
import { AuthLoginDto } from './dto/auth-login.dto'
import { AuthRegisterDto } from './dto/auth-register.dto'
import { AuthLoginResponse } from './response/auth-login.response'
import { TokenPayload } from './types/token-payload'

@Injectable()
export class AuthService {
  constructor(private jwtService: JwtService, private userService: UsersService) {}

  async login({ username, password }: AuthLoginDto): Promise<AuthLoginResponse> {
    const auth = await this.userService.getAuthenticated({
      username,
      password,
    })
    return await this._createAuthResponse(auth)
  }

  // TODO: Add a route
  async authRefresh(tokenData: any): Promise<AuthLoginResponse> {
    const user = tokenData.user
    const payload: TokenPayload = { sub: user.id, version: user.version }
    const access_token = await this.jwtService.signAsync(payload)
    return {
      access_token,
      refresh_token: tokenData.value,
      token_type: 'Bearer',
    }
  }

  async register({ username, password }: AuthRegisterDto): Promise<AuthLoginResponse> {
    const user = await this.userService.create({ username, password, role: Role.USER })
    user.version = user.version ?? 0
    return await this._createAuthResponse(user)
  }

  private async _createAuthResponse(auth: User): Promise<AuthLoginResponse> {
    const payload: TokenPayload = { sub: auth.id, version: auth.version! }
    const access_token = await this.jwtService.signAsync(payload)
    const refresh_token = await this.jwtService.signAsync(payload, { expiresIn: '7d' })
    return {
      access_token,
      refresh_token,
      token_type: 'Bearer',
    }
  }
}

File: backend/python-fastapi/server/services/auth_service.py
from fastapi import Depends

from server.dto.auth_dto import AuthLoginBody, AuthRegisterBody, AuthResponse
from server.dto.users_dto import UserCreateBody
from server.models import UserRole
from server.services.jwt_service import JwtService, TokenData, get_jwt_service
from server.services.users_service import UsersService, get_users_service


class AuthService:
    def __init__(self, user_service: UsersService,
                 jwt_service: JwtService) -> None:
        self.user_service = user_service
        self.jwt_service = jwt_service

    async def login(self, body: AuthLoginBody) -> AuthResponse:
        user = await self.user_service.get_authenticated(
            username=body.username,
            password=body.password
        )
        return self._create_auth_response(user.id, user.version)

    async def auth_refresh(self, refresh_token: TokenData) -> AuthResponse:
        user = refresh_token.user
        data = {'sub': str(user.id), 'version': user.version}
        access_token = self.jwt_service.create_access_token(data)
        return AuthResponse(access_token=access_token,
                            refresh_token=refresh_token.value)

    async def register(self, body: AuthRegisterBody) -> AuthResponse:
        user = await self.user_service.create(UserCreateBody(
            username=body.username,
            password=body.password,
            role=UserRole.USER
        ))
        return self._create_auth_response(user.id, user.version)

    def _create_auth_response(self,
                              user_id: str,
                              version: int) -> AuthResponse:
        data = {'sub': user_id, 'version': version}
        access_token = self.jwt_service.create_access_token(data)
        refresh_token = self.jwt_service.create_refresh_token(data)
        return AuthResponse(access_token=access_token,
                            refresh_token=refresh_token)


def get_auth_service(
    user_service: UsersService = Depends(get_users_service),
    jwt_service: JwtService = Depends(get_jwt_service),
) -> AuthService:
    return AuthService(user_service, jwt_service)
